interop-tests/src/test/scala/akka/grpc/scaladsl/PowerApiSpec.scala

```diff
@@ -84,5 +115,45 @@ class PowerApiSpec(backend: String)
       response.headers.getText("baz").get should be("qux")
       response.trailers.futureValue.getText("foo").get should be("bar")
     }
+
+    "(on streamed calls) redeem the headers future as soon as they're available (and trailers future when trailers arrive)" in {
+      // invoking streamed calls using the power API materializes a Future[GrpcResponseMetadata]
+      // that should redeem as soon as the HEADERS is consumed. Then, the GrpcResponseMetadata instance
+      // contains another Future that will redeem when receiving the trailers.
+      client = GreeterServiceClient(
+        GrpcClientSettings.connectToServiceAt("localhost", server.localAddress.getPort).withTls(false))
+
+      val request = HelloRequest("Alice")
+      val responseSource: Source[HelloReply, Future[GrpcResponseMetadata]] =
+        client.itKeepsReplying().invokeWithMetadata(request)
+      val firstItem: CountDownLatch = new CountDownLatch(1)
+      val fiveItems: CountDownLatch = new CountDownLatch(5)
+      val headers: Future[GrpcResponseMetadata] =
+        responseSource
+          .map { x =>
+            firstItem.countDown()
+            fiveItems.countDown()
+            x
+          }
+          .to(Sink.ignore)
+          .run
+
+      headers.isCompleted should be(false)
+
+      // As soon as the first item has arrived, the headers future
+      // should be redeemed but the fiveItems latch should still be open
+      // and the trailers future should not have been redeemed yet.
+      firstItem.await(500, TimeUnit.MILLISECONDS)
+      val trailers = headers.futureValue.trailers
+      fiveItems.getCount.toInt should be < 5
```

Comment:
I guess this is a race condition as well, as the items can arrive all at once:

```
should (on streamed calls) redeem the headers future as soon as they're available (and trailers future when trailers arrive) *** FAILED *** (180 milliseconds)
[info]   5 was not less than 5 (PowerApiSpec.scala:148)
```

I'm not sure I see a 'quick fix' other than refactoring based on `streamHellos` (and sending the 'final' hello's only after validating the (leading) headers arrived)

Reply:
> the items can arrive all at once

Ah, of course. I assumed throttling the items on the server would have been enough.

Making the server reply forever until a new message is sent from the client seems the best move forward. 👍🏼